File: patch_tool.py
from pathlib import Path
import difflib
from check_file_permission import check_file_permission

def patch_file(file_path, patch_content, config):
    if not check_file_permission(file_path, "write", config):
        raise PermissionError(f"Write access denied for {file_path}")
    file_path = Path(file_path).resolve()
    if not file_path.is_file():
        raise ValueError(f"File not found: {file_path}")
    with open(file_path, "r", encoding="utf-8") as f:
        lines = f.readlines()
    patched_lines = difflib.restore(patch_content.splitlines(), 1)
    with open(file_path, "w", encoding="utf-8") as f:
        f.writelines(patched_lines)
    return f"Applied patch to {file_path}"

File: ts-tools/patch_tool.ts

import * as fs from 'fs';
import * as path from 'path';
import * as diff from 'diff';
import { checkFilePermission } from './check_file_permission';

export function patchFile(filePath: string, patchContent: string, config: any): Promise<string> {
    return new Promise((resolve, reject) => {
        checkFilePermission(filePath, "write", config)
            .then(() => {
                const resolvedPath = path.resolve(filePath);
                fs.readFile(resolvedPath, 'utf-8', (err, data) => {
                    if (err) {
                        reject(err);
                    } else {
                        const patched = diff.applyPatch(data, patchContent);
                        if (patched === false) {
                            reject(new Error("Failed to apply patch"));
                            return;
                        }
                        fs.writeFile(resolvedPath, patched, 'utf-8', (err) => {
                            if (err) {
                                reject(err);
                            } else {
                                resolve(`Applied patch to ${filePath}`);
                            }
                        });
                    }
                });
            })
            .catch(reject);
    });
}
